examples/tv-app/tv-common/clusters/messages/MessagesManager.cpp

```diff
@@ -18,62 +18,143 @@
 
 // Commands
-void MessagesManager::HandlePresentMessagesRequest(
-    const chip::ByteSpan & messageId, const MessagePriorityEnum & priority,
-    const chip::BitMask<MessageControlBitmap> & messageControl, const chip::app::DataModel::Nullable<uint32_t> & startTime,
-    const chip::app::DataModel::Nullable<uint16_t> & duration, const chip::CharSpan & messageText,
-    const chip::Optional<chip::app::DataModel::DecodableList<MessageResponseOption>> & responses)
+void MessagesManager::HandlePresentMessagesRequest(const ByteSpan & messageId, const MessagePriorityEnum & priority,
+                                                   const BitMask<MessageControlBitmap> & messageControl,
+                                                   const DataModel::Nullable<uint32_t> & startTime,
+                                                   const DataModel::Nullable<uint16_t> & duration, const CharSpan & messageText,
+                                                   const Optional<DataModel::DecodableList<MessageResponseOption>> & responses)
 {
-    Message message{
-        // TODO: Enable id
-        chip::ByteSpan(), priority, messageControl, startTime, duration,
-        // TODO: Enable text
-        chip::CharSpan()
-        // TODO: Convert responses to Optional<chip::app::DataModel::List<const
-        // chip::app::Clusters::Messages::Structs::MessageResponseOptionStruct::Type>> message.responses = responses;
-    };
-
-    mMessages.push_back(message);
+    std::unique_ptr<char[]> messageTextBufferPtr;
+    std::unique_ptr<uint8_t[]> messageIdBufferPtr;
+
+    uint8_t * messageIdBuffer = nullptr;
+    // see MessagesManager.h "Note on memory management" for allocation/free rules
+    char * messageTextBuffer = new char[messageText.size() + 1];
+    VerifyOrExit(messageTextBuffer != nullptr,
+                 ChipLogProgress(Controller, "HandlePresentMessagesRequest messageTextBuffer alloc failed"));
+    messageTextBufferPtr.reset(messageTextBuffer);
+
+    memcpy(messageTextBuffer, messageText.data(), messageText.size());
+    messageTextBuffer[messageText.size()] = '\0';
+
+    ChipLogProgress(Controller, "HandlePresentMessagesRequest message:%s size:%lu", messageTextBuffer, messageText.size());
+
+    // see MessagesManager.h "Note on memory management" for allocation/free rules
+    messageIdBuffer = new uint8_t[messageId.size()];
+    VerifyOrExit(messageIdBuffer != nullptr,
+                 ChipLogProgress(Controller, "HandlePresentMessagesRequest messageIdBuffer alloc failed"));
+    memcpy(messageIdBuffer, messageId.data(), messageId.size());
+    messageIdBufferPtr.reset(messageIdBuffer);
+
+    {
+        messageTextBufferPtr.release();
+        messageIdBufferPtr.release();
+        CachedMessage * cachedMessage = new CachedMessage(messageIdBuffer, messageId.size(), priority, messageControl, startTime,
+                                                          duration, std::string(messageTextBuffer, messageText.size() + 1));
```

Comment:
This leaks since you released messageTextBufferPtr and std::string does not take ownership.

Reply:
moved memory management to CachedMessage and CachedMessageOption